tsdb/engine/tsm1/cache.go

```diff
@@ -180,37 +187,55 @@ func (c *Cache) Keys() []string {
 
 // Values returns a copy of all values, deduped and sorted, for the given key.
 func (c *Cache) Values(key string) Values {
-	values, needSort := func() (Values, bool) {
-		c.mu.RLock()
-		defer c.mu.RUnlock()
-		e := c.store[key]
-		if e == nil {
-			return nil, false
+	c.mu.Lock()
```

Comment:
There is 1 particular corner case that makes it tricky. If the system starts with a read-lock, sees that sort is required, and moves to a write lock, it may need to then keep the write-lock for the remainder of the function. This is because if the write lock was released immediately after the sorting was performed, `Snapshot` could run and move the entries currently in the hot cache into the snapshot-caches. This would mean that data that has just been sorted (and copied) is now also in the snapshot-cache. So now you have to check for that case or your data will be doubled (yes, you could try to dedupe I guess).

Keeping the write-lock, once taken, for the rest of the function avoids the need to check, since the `Snapshot` cannot run. Perhaps that is OK, since most of the time the data for a given key *will* be sorted, and the write-lock will usually not be taken.